source/common/upstream/conn_pool_map_impl.h

```diff
@@ -6,30 +6,54 @@ namespace Envoy {
 namespace Upstream {
 
 template <typename KEY_TYPE, typename POOL_TYPE>
-ConnPoolMap<KEY_TYPE, POOL_TYPE>::ConnPoolMap(Envoy::Event::Dispatcher& dispatcher)
-    : thread_local_dispatcher_(dispatcher) {}
+ConnPoolMap<KEY_TYPE, POOL_TYPE>::ConnPoolMap(Envoy::Event::Dispatcher& dispatcher,
+                                              absl::optional<uint64_t> max_size)
+    : thread_local_dispatcher_(dispatcher), max_size_(max_size) {}
 
 template <typename KEY_TYPE, typename POOL_TYPE>
 ConnPoolMap<KEY_TYPE, POOL_TYPE>::~ConnPoolMap() = default;
 
 template <typename KEY_TYPE, typename POOL_TYPE>
-POOL_TYPE& ConnPoolMap<KEY_TYPE, POOL_TYPE>::getPool(KEY_TYPE key, const PoolFactory& factory) {
+typename ConnPoolMap<KEY_TYPE, POOL_TYPE>::OptPoolRef
+ConnPoolMap<KEY_TYPE, POOL_TYPE>::getPool(KEY_TYPE key, const PoolFactory& factory) {
   Common::AutoDebugRecursionChecker assert_not_in(recursion_checker_);
   // TODO(klarose): Consider how we will change the connection pool's configuration in the future.
   // The plan is to change the downstream socket options... We may want to take those as a parameter
   // here. Maybe we'll pass them to the factory function?
-  auto inserted = active_pools_.emplace(key, nullptr);
-
-  // If we inserted a new element, create a pool and assign it to the iterator. Tell it about any
-  // cached callbacks.
-  if (inserted.second) {
-    inserted.first->second = factory();
-    for (const auto& cb : cached_callbacks_) {
-      inserted.first->second->addDrainedCallback(cb);
+  auto pool_iter = active_pools_.find(key);
+  if (pool_iter != active_pools_.end()) {
+    return std::ref(*(pool_iter->second));
+  }
+
+  // We need a new pool. Check if we have room.
+  if (max_size_.has_value() && size() >= max_size_.value()) {
+    // We have hit the maximum number of connection pools. Ask the tracked pools to inform us when
+    // we can free them.
+    registerFullCallbacks();
```

Comment:
I think I better understand what is going on here now. Thanks for the comments. Q: Can all of this be simplified by doing the following: Loop over all pools, and instead of registering callbacks, just have some method called something like `hasActiveConnections()` or something like that, and if not, just delete it inline? I think if we did this, a lot of things become a lot simpler, we just get rid of the first one we find, no need for another vector, etc. and also then we don't leave callbacks installed after we do this operation which seems not optimal and potentially will lead to strange edge case problems?

Reply:
That sounds like a good approach. We'll have the same semantics -- we know right away if there's one free, and if there isn't we'll fail. If one frees up later, then the next request will work.

The only downside I can see to this is that if there are a large number of connection pools, and we're at the limit with pools only freeing up occasionally, we'll end up O(n) iterating over them for each new request. This could be slow for a large limit. Of course, the current code I have up doesn't really solve this for that case either. My original design here had us tracking the "idle" state of a pool so we could have a ready-to-go list of pools which could be freed.

For this first round, we definitely don't need that. I'd  rather build the rest of the solution out and then optimize it.